addon-test-support/@ember/test-helpers/-internal/warnings.ts

```diff
@@ -0,0 +1,109 @@
+import { BaseContext } from '../setup-context';
+import { registerWarnHandler } from '@ember/debug';
+
+export interface WarningOptions {
+  id?: string;
+}
+
+export interface Warning {
+  message: string;
+  options?: WarningOptions;
+}
+
+// the WARNINGS data structure which is used to weakly associated warnings with
+// the test context their occured within
+const WARNINGS = new WeakMap<BaseContext, Array<Warning>>();
+
+/**
+ *
+ * Provides the list of warnings associated with a given base context;
+ *
+ * @private
+ * @param {BaseContext} [context] the test context
+ * @return {Array<Warning>} the warnings associated with the corresponding BaseContext;
+ */
+export function getWarningsForContext(context: BaseContext): Array<Warning> {
+  if (!context) {
+    throw new TypeError(
+      `[@ember/test-helpers] could not get warnings for an invalid test context: '${context}'`
+    );
+  }
+
+  let warnings = WARNINGS.get(context);
+
+  if (!Array.isArray(warnings)) {
+    warnings = [];
+    WARNINGS.set(context, warnings);
+  }
+
+  return warnings;
+}
+
+/**
+ *
+ * Provides the list of warnings associated with a given test context which
+ * occured only while a the provided callback is executed. This callback can be
+ * synchonous, or it can be an async function.
+ *
+ * @private
+ * @param {BaseContext} [context] the test context
+ * @param {CallableFunction} [callback] The callback that when executed will have its warnings recorded
+ * @return {Array<Warning>} The warnings associated with the corresponding baseContext which occured while the CallbackFunction was executed
+ */
+export function getWarningsDuringCallbackForContext(
+  context: BaseContext,
+  callback: CallableFunction
+): Array<Warning> | Promise<Array<Warning>> {
+  if (!context) {
+    throw new TypeError(
+      `[@ember/test-helpers] could not get warnings for an invalid test context: '${context}'`
+    );
+  }
+
+  const warnings = getWarningsForContext(context);
+  const previousLength = warnings.length;
+
+  const result = callback();
+
+  if (
+    (result !== null && typeof result === 'object') ||
+    (typeof result === 'function' && typeof result.then === 'function')
+  ) {
```

Comment:
Mind extracting this to 

https://github.com/emberjs/ember-test-helpers/blob/master/addon-test-support/%40ember/test-helpers/-utils.ts

It would be much easier to read as `if (isPromise(result)) {`

Reply:
done